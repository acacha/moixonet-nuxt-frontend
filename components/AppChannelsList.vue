<template>
  <ul>
    <li v-for="channel in channels" :key="channel.id">{{ channel.name }}</li>
  </ul>
</template>

<script>
export default {
  name: 'VChannelsList',
  props: {
    channels: {
      type: Array,
      required: true
    }
  }
}
</script>
