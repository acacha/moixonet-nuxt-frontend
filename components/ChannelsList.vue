<template>
  <v-container>
    <v-card>
      <v-toolbar dense color="primary">
        <v-toolbar-title class="white--text">Llista de canals</v-toolbar-title>
        <v-spacer></v-spacer>

        <v-toolbar-items>
          <v-btn text icon class="white--text" @click="refresh" :loading="loading">
            <v-icon>mdi-cached</v-icon>
          </v-btn>
        </v-toolbar-items>

      </v-toolbar>
      <v-list>
        <v-list-item v-for="channel in channels">
          {{ channel.name }}
        </v-list-item>
        <v-card-actions>
          <v-btn @click="omplir" color="primary">
            Omplir
          </v-btn>
          <v-btn @click="buidar" color="error">
            Buidar
          </v-btn>
          <v-btn @click="refresh">
            refresh
          </v-btn>
        </v-card-actions>
      </v-list>
    </v-card>
  </v-container>
</template>

<script>
// import axios from 'axios'
import channelsFixture from '../cypress/fixtures/channels'
export default {
  name: 'ChannelsList',
  data () {
    return {
      loading: false
    }
  },
  computed: {
    channels () {
      // return []
      return this.$store.getters['channels/channels']
    }
  },
  created () {
    this.refresh()
  },
  methods: {
    omplir () {
      // MODIFICAR L'ESTAT
      this.$store.commit('channels/SET_CHANNELS', channelsFixture)
    },
    buidar () {
      // MODIFICAR L'ESTAT
      this.$store.commit('channels/SET_CHANNELS', [])
    },
    async refresh () {
      this.loading = true
      await this.$store.dispatch('channels/REFRESH')
      this.loading = false
    },
    async refreshold () {
      try {
        const channels = await this.$axios.get('http://moixonet-backend.test/api/v1/channels')
        this.$store.commit('channels/SET_CHANNELS', channels.data)
      } catch (error) {
        console.log(error)
      }
    },
    refreshPromise () {
      this.$axios.get('http://moixonet-backend.test/api/v1/channels').then((channels) => {
        this.$store.commit('channels/SET_CHANNELS', channels.data)
      }).catch((error) => {
        console.log(error)
      })
    }
  }
}
</script>
