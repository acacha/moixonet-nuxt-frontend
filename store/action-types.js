export const CHANNEL_STORE = 'CHANNEL_STORE'
export const CHANNEL_DESTROY = 'CHANNEL_DESTROY'
export const CHANNELS_INDEX = 'CHANNELS_INDEX'
export const CHANNEL_UPDATE = 'CHANNEL_UPDATE'
