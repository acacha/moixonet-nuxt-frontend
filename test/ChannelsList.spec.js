import { shallowMount } from '@vue/test-utils'
import ChannelsList from '../components/ChannelsList'
import AppChannelsList from '../components/AppChannelsList'
import initialState from '../store/channels/state'
import channelsFixture from '../cypress/fixtures/channels'
import * as actionTypes from '../action-types'

describe('ChannelsList', () => {
  let opts
  const actions = jest.fn()
  const optsFixture = {
    mocks: {
      $store: {
        state: {
          channels: {
            list: channelsFixture
          }
        },
        dispatch: actions
      }
    }
  }

  const build = (options) => {
    options = options || opts
    const wrapperShallow = shallowMount(ChannelsList, options)

    return {
      wrapperShallow,
      channelsList: () => wrapperShallow.find(AppChannelsList)
    }
  }

  beforeEach(() => {
    // MOCKING this.$store.state.channels.list
    opts = {
      mocks: {
        $store: {
          state: {
            channels: initialState()
          }
        }
      }
    }
  })

  test('is a Vue instance', () => {
    const { wrapperShallow } = build()
    expect(wrapperShallow.isVueInstance()).toBeTruthy()
  })

  test('renders the component with no channels', () => {
    const { wrapperShallow } = build()
    expect(wrapperShallow.element).toMatchSnapshot()
  })

  test('renders the component channels', () => {
    const { wrapperShallow } = build(optsFixture)
    expect(wrapperShallow.element).toMatchSnapshot()
  })

  test('renders child components', () => {
    const { channelsList } = build()
    expect(channelsList().exists()).toBe(true)
  })

  test('Passes a channels prop to channels list component', () => {
    const { channelsList } = build(optsFixture)
    expect(channelsList().vm.channels).toBe(channelsFixture)
  })

  test('Removes a channel when received "removed"', () => {
    const { channelsList } = build(optsFixture)
    const removedChannel = channelsFixture[0]

    channelsList().vm.$emit('removed', removedChannel)

    expect(actions).toHaveBeenCalled()
    expect(actions.mock.calls[0][0]).toEqual('channels/' + actionTypes.CHANNEL_DESTROY)
    expect(actions.mock.calls[0][1]).toEqual(removedChannel)
  })
})
