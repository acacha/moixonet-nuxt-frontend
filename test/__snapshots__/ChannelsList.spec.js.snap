// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppChannelsList renders a list of channels 1`] = `
<ul>
  <li>
    Canal 1
  </li>
  <li>
    Canal 2
  </li>
  <li>
    Canal 3
  </li>
</ul>
`;

exports[`ChannelsList renders a list of channels 1`] = `
<ul>
  <li>
    Canal 1
  </li>
  <li>
    Canal 2
  </li>
  <li>
    Canal 3
  </li>
</ul>
`;

exports[`ChannelsList renders and html list 1`] = `
<ul>
  <li>
    Canal 1
  </li>
  <li>
    Canal 2
  </li>
  <li>
    Canal 3
  </li>
</ul>
`;
